CaloReco/src/FastCaloRecoDigiFromDigi_module.cc

```diff
@@ -0,0 +1,142 @@
+#include "RecoDataProducts/inc/CaloRecoDigiCollection.hh"
+
+#include "DataProducts/inc/EventWindowMarker.hh"
+
+#include <iostream>
+#include <string>
+
+
+#include <iostream>
+#include <string>
+   
+namespace mu2e {
+
+  class FastCaloRecoDigiFromDigi : public art::EDProducer {
+
+  	public:
+		explicit FastCaloRecoDigiFromDigi(fhicl::ParameterSet const& pset) :
+		art::EDProducer{pset}, caloDigisToken_{consumes<CaloDigiCollection>(pset.get<std::string>("caloDigiModuleLabel"))},
+		digiSampling_       (pset.get<double>("digiSampling")),
+		minDigiE_   	    (pset.get<double> ("minDigiE")),
+		shiftTime_          (pset.get<double> ("shiftTime")),
+		diagLevel_          (pset.get<int>    ("diagLevel",0)),
+		ewMarkerTag_(pset.get<art::InputTag>("EventWindowMarkerLabel"))
+	     	{
+			consumes<EventWindowMarker>(ewMarkerTag_);
+	      		produces<CaloRecoDigiCollection>();
+		}
+
+    		void produce(art::Event& e) override;
+
+	private:
+
+		art::ProductToken<CaloDigiCollection> const caloDigisToken_;
+		double const digiSampling_;
+		double 	 minDigiE_ ;
+		double 	 shiftTime_ ;
+		int      diagLevel_;
+   		art::InputTag ewMarkerTag_;// name of the module that makes eventwindowmarkers
+   
+		void extractRecoDigi(art::ValidHandle<CaloDigiCollection> const& caloDigis, CaloRecoDigiCollection& recoCaloHits, double const& ewOffset);
+
+	};
+
+	void FastCaloRecoDigiFromDigi::produce(art::Event& event)
+  	{
+
+		if (diagLevel_ > 0) std::cout<<"[FastRecoDigiFromDigi::produce] begin"<<std::endl;
+
+		auto const& caloDigisH = event.getValidHandle(caloDigisToken_);
+		auto recoCaloDigiColl = std::make_unique<CaloRecoDigiCollection>();
+		
+		double ewmOffset = 0;
+		art::Handle<EventWindowMarker> ewMarkerHandle;
+		if (event.getByLabel(ewMarkerTag_, ewMarkerHandle)){
+			const EventWindowMarker& ewMarker(*ewMarkerHandle);
+			ewmOffset = ewMarker.timeOffset();
+		}
+
+		extractRecoDigi(caloDigisH, *recoCaloDigiColl, ewmOffset);
+
+		if ( diagLevel_ > 0 )std::cout<<"[FastCaloRecoDigiFromDigi::produce] extracted "<<recoCaloDigiColl->size()<<" RecoDigis"<<std::endl;
+		
+
+		event.put(std::move(recoCaloDigiColl));
+
+		if (diagLevel_ > 0) std::cout<<"[FastRecoDigiFromDigi::produce] end"<<std::endl;
+
+		return;
+	}
+
+	void FastCaloRecoDigiFromDigi::extractRecoDigi(art::ValidHandle<CaloDigiCollection> const& caloDigisHandle, CaloRecoDigiCollection &recoCaloHits, double const& eventWindowOffset)
+  	{
+
+		std::vector<double> x,y;
+		ConditionsHandle<CalorimeterCalibrations> calorimeterCalibrations("ignored");
+
+		auto const& caloDigis = *caloDigisHandle;
+		
+		CaloDigi const* base = &caloDigis.front(); 
+
+		for (const auto& caloDigi : caloDigis)
+		{
+    
+			int    roId     = caloDigi.roId();
+			double t0       = caloDigi.t0();
+			double adc2MeV  = calorimeterCalibrations->Peak2MeV(roId);
+			const std::vector<int>& waveform = caloDigi.waveform();
+
+			size_t index = &caloDigi - base;
+			art::Ptr<CaloDigi> caloDigiPtr(caloDigisHandle, index);
+
```

Comment:
Do we need to fill the x,y vectors here? Could we do instead
time = t0 + (caloDigi.peakpos()+0.5)*digiSampling_+eventWindowOffset;
eDep = waveform.at(caloDigi.peakpos())*adc2MeV;

and avoid those two vectors. They don't seem to be used anywhere else.

Reply:
I agree. Probably, in the t0 formula, we should already add a  "+calorimeterCalibrations->timeOffset(roId)" to remind us to add the calibrated channel time offset?

Reply:
yes, just replying to this. I will test but I think you are correct

Reply:
done